src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -0,0 +1,298 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Packaging;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.Shell.Interop;
+using NuGet.VisualStudio;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Packaging
+{
+    [ExportWorkspaceServiceFactory(typeof(IPackageInstallerService)), Shared]
+    internal class PackageInstallerServiceFactory : IWorkspaceServiceFactory
+    {
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            return new PackageInstallerService();
+        }
+    }
+
+    internal class PackageInstallerService : ForegroundThreadAffinitizedObject, IPackageInstallerService
+    {
+        private readonly object _gate = new object();
+
+        private VisualStudioWorkspaceImpl _workspace;
+        private IVsPackageInstallerServices _packageInstallerServices;
+        private IVsPackageInstaller _packageInstaller;
+
+        private CancellationTokenSource _tokenSource = new CancellationTokenSource();
+        private bool _solutionChanged;
```

Comment:
Add a comment to explain that we find out about solution changes at a point when we don't actually have access to the solution on the UI thread, and so this is used to communicate that state to the later processing that happens on the UI thread (otherwise, we would just calculate this up front and not need to save it off like this).